### PR TITLE
[FLINK-21794][metrics] Support retrieving slot details via rest api

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ClusterResourceStatisticsProvider.java [hunk 1]
```diff
@@ -50,4 +52,7 @@
 
     /** Get profile of total pending resources. */
     ResourceProfile getPendingResource();
+
+    /** Get allocate slots information from the TaskManager with the given instance id. */
+    Collection<TaskManagerSlotInformation> getAllocatedSlotsOf(InstanceID instanceId);
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/ClusterResourceStatisticsProvider.java:57]: This is not **statistics**.

I'm not entirely sure whether we need to add this interface. Given that it is only used in `FineGrainedSlotManager#getAllocatedSlotsOf`, we may consider to move the implementation of `FineGrainedTaskManagerTracker#getAllocatedSlotsOf` to `FineGrainedSlotManager`.

I'm also fine with adding `getAllocatedSlotsOf` as an shortcut interface. In that case, we might consider to 1) move it to `TaskManagerResourceInfoProvider`, and 2) provide a default implementation basing on `getRegisteredTaskManager`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java [hunk 2]
```diff
@@ -717,6 +717,12 @@ public ResourceProfile getFreeResourceOf(InstanceID instanceID) {
         return taskExecutorManager.getTotalFreeResourcesOf(instanceID);
     }
 
+    @Override
+    public Collection<TaskManagerSlotInformation> getAllocatedSlotsOf(InstanceID instanceID) {
+        // This information is only valid for fine-grained resource management.
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java:722]: I would say it's currently not supported for this slot manager. The information should always be valid.

Same for `SlotManagerImpl`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java [hunk 3]
```diff
@@ -70,6 +71,8 @@
 
     ResourceProfile getFreeResourceOf(InstanceID instanceID);
 
+    Collection<TaskManagerSlotInformation> getAllocatedSlotsOf(InstanceID instanceID);
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java:74]: This interface exposes `TaskManagerSlotInformation`, which should be internal to slot manager.
I'd suggest to return `SlotInfo` here.

**Review by xintongsong — APPROVED** [review]
Thanks @KarmaGYZ, LGTM.
@vthinkxie, could you help double-check on this PR?

**Review by vthinkxie — APPROVED** [review]
thanks @KarmaGYZ 
LGTM

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerInfo.java [hunk 4]
```diff
@@ -62,6 +63,8 @@
 
     public static final String FIELD_NAME_MEMORY = "memoryConfiguration";
 
+    public static final String FIELD_NAME_ALLOCATE_SLOTS = "allocateSlots";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerInfo.java:66]:
```suggestion
    public static final String FIELD_NAME_ALLOCATE_SLOTS = "allocatedSlots";
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java [hunk 5]
```diff
@@ -636,6 +637,15 @@ public ResourceProfile getFreeResourceOf(InstanceID instanceID) {
         return taskManagerTracker.getRegisteredResourceOf(instanceID);
     }
 
+    @Override
+    public Collection<SlotInfo> getAllocatedSlotsOf(InstanceID instanceID) {
+        return taskManagerTracker.getRegisteredTaskManager(instanceID)
+                .map(TaskManagerInfo::getAllocatedSlots).map(Map::values)
+                .orElse(Collections.emptyList()).stream()
+                .map(slot -> new SlotInfo(slot.getJobId(), slot.getResourceProfile()))
+                .collect(Collectors.toList());
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:646]: This will scale terribly. Do we really need information for every single slot?
Why not slots + number of slots of that type? Or total acquired resources? Or available/total resources of the TM?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/SlotInfo.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        this(jobId, ResourceProfileInfo.fromResrouceProfile(resource));
+    }
+
+    public JobID getJobId() {
+        return jobId;
+    }
+
+    public ResourceProfileInfo getResource() {
+        return resource;
+    }
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/SlotInfo.java:74]: getters should be annotated with `@JsonIgnore` to prevent accidends when fields are renamed

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java (hunk 5 above)
Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:646]: Currently, we already have the available/total resources of the TM. For fine-grained resource management, slots are dynamically created with different resource profiles. It would be good to let user know the what type of slot is allocated and which job it is allocated for. I don't think it will help allocate to reorganize it to slots + number of slots of that type. In the worst case, that will increase the data size instead.

After an offline discussion with @vthinkxie , I think we can add only this information to `TaskManagerDetailsInfo`.

**Comment by KarmaGYZ** [comment]
Thanks for the comments @zentol . PR updated.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java (hunk 5 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:646]: But why bother with slots in the first place? Isn't the whole point for fine-grained resource management to not really bother with slots but _resources_ instead, so why then expose it as slots again in the UI?

Why not just say "Job A has allocated X memory and Y CPUs on TM Z to run N tasks."

What value does it bring to know this instead:
"Job as has N slots on TM Z, where slot 1 uses X_1 memory any Y_1 cpus, and slot 2 uses X_2 memory and Y_2 cpus, and ..."

Comment (vthinkxie) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:646]: "Job as has N slots on TM Z, where slot 1 uses X_1 memory any Y_1 cpus, and slot 2 uses X_2 memory and Y_2 cpus, and ..." could be aggregated and get "Job A has allocated X memory and Y CPUs on TM Z to run N tasks."
while it cannot be pushed backward.
Since TM and resources are implemented on a slots basis, the interface layer should expose atomic properties

Comment (KarmaGYZ) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManager.java:646]: I think how to expose that information to the user is depends on the WebUI and might be out of the scope of this PR. In current interface, we give the UI flexibility to choose how to aggregate it.
Also, user might want to know the explicit resource of each task, especially for external resources.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerDetailsHandler.java [hunk 7]
```diff
@@ -87,14 +87,14 @@ public TaskManagerDetailsHandler(
         final ResourceID taskManagerResourceId =
                 request.getPathParameter(TaskManagerIdPathParameter.class);
 
-        CompletableFuture<TaskManagerInfo> taskManagerInfoFuture =
-                gateway.requestTaskManagerInfo(taskManagerResourceId, timeout);
+        CompletableFuture<TaskManagerInfoWithSlots> taskManagerInfoWithSLotsFuture =
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/TaskManagerDetailsHandler.java:90]:
```suggestion
        CompletableFuture<TaskManagerInfoWithSlots> taskManagerInfoWithSlotsFuture =
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerDetailsInfo.java [hunk 8]
```diff
@@ -70,31 +80,39 @@ public TaskManagerDetailsInfo(
...
+                taskManagerInfoWithSlots.getTaskManagerInfo().getTotalResource(),
+                taskManagerInfoWithSlots.getTaskManagerInfo().getFreeResource(),
+                taskManagerInfoWithSlots.getTaskManagerInfo().getHardwareDescription(),
+                taskManagerInfoWithSlots.getTaskManagerInfo().getMemoryConfiguration(),
+                taskManagerInfoWithSlots.getAllocatedSlots(),
                 taskManagerMetrics);
     }
 
-    @JsonProperty(FIELD_NAME_METRICS)
+    @JsonIgnore
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerDetailsInfo.java:105]: This is a bit unrelated. We may consider to have this in a hotfix commit, together with properties in `TaskManagerInfo`.

**Review by xintongsong — COMMENTED** [review]
LGTM.
@zentol, would you like to have another look at this?
